File: src/Models/GameSessionHelpers.swift
import Foundation

// Shared 1-second repeating timer used by every game's ViewModel to drive `state.timerSeconds`.
// The ViewModel still owns `state.isTimerActive`/`state.timerSeconds` (they're part of each
// game's persisted/undoable state); this just wraps the Timer scheduling that was previously
// copy-pasted identically into GameViewModel, BeecellViewModel, and SpiderViewModel.
final class GameTimer {
    private var timer: Timer?

    func start(isActive: () -> Bool, setActive: (Bool) -> Void, tick: @escaping () -> Void) {
        guard !isActive() else { return }
        setActive(true)
        timer = Timer.scheduledTimer(withTimeInterval: 1.0, repeats: true) { _ in tick() }
    }

    func stop(setActive: (Bool) -> Void) {
        timer?.invalidate()
        timer = nil
        setActive(false)
    }
}

// Bounded undo history shared by every game's ViewModel — caps at `capacity` snapshots,
// dropping the oldest once full.
struct UndoStack<State> {
    private var stack: [State] = []
    private let capacity: Int

    init(capacity: Int = 100) {
        self.capacity = capacity
    }

    var isEmpty: Bool { stack.isEmpty }

    mutating func push(_ state: State) {
        stack.append(state)
        if stack.count > capacity {
            stack.removeFirst()
        }
    }

    mutating func pop() -> State? {
        stack.isEmpty ? nil : stack.removeLast()
    }

    mutating func removeAll() {
        stack.removeAll()
    }
}

// Smart Drop Detection: when a player drags a stack, the tapped card might not be the one
// they meant to grab (e.g. grabbing a 10♠ sitting just above the 9♥ they actually wanted).
// Rather than rejecting the whole drop, this finds the longest suffix of the dragged stack —
// trimming cards off the grabbed end, never the bottom — that is a legal move onto the
// target. Trimmed leading cards are simply left behind in the source pile (never removed),
// so nothing needs to visually "snap back": the drag overlay resets and the untouched cards
// are exactly where they always were.
enum SmartDrop {
    static func resolve(cards: [Card], isValidMove: ([Card]) -> Bool) -> [Card]? {
        guard !cards.isEmpty else { return nil }
        for start in 0..<cards.count {
            let suffix = Array(cards[start...])
            if isValidMove(suffix) {
                return suffix
            }
        }
        return nil
    }
}

// Shared win-condition check: the foundation pile count reaching the target total, and the
// game not already marked won. Each ViewModel still owns its own win-completion side effects
// (stopTimer/recordWin/playSound/score adjustments) since those differ per game.
enum WinDetection {
    static func hasWon(foundationCardCount: Int, totalCards: Int, alreadyWon: Bool) -> Bool {
        foundationCardCount == totalCards && !alreadyWon
    }
}

// Point Highlights: a transient "+N"/"-N" (or "+$0.50"/"-$5.00" in Vegas mode) popup
// flashed over the specific card responsible for a score change. Each ViewModel
// (Klondike/Beecell/Spider) holds one of these directly as a plain `var` — not part of
// its Codable state struct, same precedent as `isAutoplayRunning`/`isStuck` — and clears
// it after a short delay via a generation counter, the same stale-callback guard used
// for `aiMoveGeneration` in Honeycomb's HoneycombViewModel.
public struct CardPointPopup: Equatable {
    public let cardId: UUID
    public let displayText: String   // e.g. "+10", "-15", "+$0.50" — already formatted
    public let isPositive: Bool      // drives popup color (green vs red)

    public init(cardId: UUID, displayText: String, isPositive: Bool) {
        self.cardId = cardId
        self.displayText = displayText
        self.isPositive = isPositive
    }
}

File: android/app/src/main/java/com/leah/honeycomb/GameSessionHelpers.kt
package com.leah.honeycomb

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import java.util.UUID

class GameTimer(private val scope: CoroutineScope) {
    private var job: Job? = null

    fun start(checkActive: () -> Boolean, onSetActive: (Boolean) -> Unit, tick: () -> Unit) {
        if (checkActive()) return
        onSetActive(true)
        job = scope.launch {
            while (isActive && checkActive()) {
                delay(1000)
                tick()
            }
        }
    }

    fun stop(onSetActive: (Boolean) -> Unit) {
        job?.cancel()
        job = null
        onSetActive(false)
    }
}

object SmartDrop {
    fun resolve(cards: List<Card>, isValidMove: (List<Card>) -> Boolean): List<Card>? {
        if (cards.isEmpty()) return null
        for (start in cards.indices) {
            val suffix = cards.subList(start, cards.size)
            if (isValidMove(suffix)) {
                return suffix
            }
        }
        return null
    }
}

object WinDetection {
    fun hasWon(foundationCardCount: Int, totalCards: Int, alreadyWon: Boolean): Boolean {
        return foundationCardCount == totalCards && !alreadyWon
    }
}

data class CardPointPopup(
    val cardId: UUID,
    val displayText: String,
    val isPositive: Boolean
)
